I2C and External Interrupt Controller Drivers for Atmel's SAMD21 MCU

madlain opened the pull request:
Hi,

I've implemented a basic I2C driver and an external interrupt controller driver for Atmel's SAMD21 MCU. I've performed functional testing for both by exercising the sample application for the FXAS21002 gyroscope running on an Arduino MKRZero board connecting to Adafruits NXP Precision 9DoF breakout board.

Madani.

madlain committed:
i2c: add a basic I2C driver for the SAM0 family

Only the most basic operation has been tested. Current limitations
include:

- the use of polling
- the rudimentary handling of concurrency
- support for standard mode only

Signed-off-by: Madani Lainani <madani.lainani@gmail.com>

madlain committed:
boards: Add Arduino MKRZero

Built upon the existing SoC for Atmel's SAMD21 MCU but with support
enabled for two newly added peripherals:

- I2C port (SERCOM)
- External Interrupt Controller

Signed-off-by: Madani Lainani <madani.lainani@gmail.com>

madlain committed:
gpio: sam0: add external interrupt controller support

This extends the existing SAMD21 GPIO driver functionnality with the
possibility to configure I/O pins as interrupt lines.

Signed-off-by: Madani Lainani <madani.lainani@gmail.com>

dbkinder reviewed (approved):
+1 for doc, thanks!

dbkinder commented:
... but you do have lots of issues with coding style noted in the shippable details above :(

madlain commented:
I'm not sure why the shippable run successive to the merge conflict resolution is failing. Can you please help me out?

madlain commented:
Is there anything needed from me to move forward with the merge?

galak commented:
@anangl can you review the i2c driver.

MaureenHelm reviewed (changes requested):
Thanks for your contribution!

Will you please add the new board to `samples/sensor/fxas21002/sample.yaml`? This will ensure that the i2c driver gets built in CI.

madlain commented on boards/arm/arduino_mkrzero/arduino_mkrzero.dts:
```diff
@@ -0,0 +1,67 @@
+
+&sercom0 {
+	status = "ok";
+	compatible = "atmel,sam0-i2c";
```
I've adhered to the exact same approach taken for the arduino_zero also based on Atmel SAMD21: the SoC device tree file only declares SERCOM devices while the board device tree file specifies how a given SERCOM is configured (USART, SPI or I2C) for that board. Is there anything wrong with this approach?

madlain commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -9,10 +10,167 @@
+#define GPIO_SAM0_PORTA_EXTINT(n) \
+	{ CONFIG_GPIO_SAM0_PORTA_EXTINT_##n##_PIN, n },
+
+static struct extint porta_extints[] = {
```
The porta_extints array allows using designated initializers for both the
mapping table and its length in the const config struct. Moving the definition
of the table to the const config struct prevents using a designated initializer
for the length and defers its computation to run time, using pointer arithmetic
in the device init function.

madlain commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -113,13 +366,187 @@ static int gpio_sam0_read(struct device *dev, int access_op, u32_t pin,
+	struct gpio_sam0_data *data;
+	u32_t enabled_int;
+
+#ifdef CONFIG_GPIO_SAM0_PORTA_BASE_ADDRESS
```
Not sure how this macro could help here. This code is in the Interrupt Service
Routine for the External Interrupt Controller where the only way to get a
reference to the structure for the GPIO ports is to use their respective device
names with DEVICE_GET.

madlain commented on boards/arm/arduino_mkrzero/arduino_mkrzero_defconfig:
```diff
@@ -0,0 +1,51 @@
+# CONFIG_FXAS21002_RANGE=3
+
+# CONFIG_FXAS21002_GPIO_NAME="PORTB"
+# CONFIG_FXAS21002_GPIO_PIN=2
```
The board doesn't carry an gyro so I've removed  (commented out) all configs including the range. The reason why these were present is that, **at the time**, it was possible to build the sample app for the sensor with an **external gyro board** hooked up to the MKRZero. This set-up provided a convenient way to exercise the newly developed I2C and EIC drivers all together with the already existing drivers for the SoC (UART, GPIO, PINMUX).

madlain commented on boards/arm/arduino_mkrzero/doc/arduino_mkrzero.rst:
```diff
@@ -0,0 +1,201 @@
++-----------+------------+--------------------------------------+
+| USART     | on-chip    | Serial ports                         |
++-----------+------------+--------------------------------------+
+| SPI       | on-chip    | Serial Peripheral Interface ports    |
```
I've enabled and tested the USART, GPIO, and PINMUX drivers.

madlain commented:
Hi,

Thanks for the review and the valuable comments. I’ve addressed and resolved all but three of them for which I’m expecting your feedback. Also as clarification, and to answer your top-level comment, the board doesn't carry the FXAS21002 gyro. At the time I implemented the I2C and EIC drivers, it was possible to build the sample app for the sensor with an external gyro board hooked up to the MKRZero. This set-up provided a convenient way to exercise the newly developed I2C and EIC drivers all together with the already existing drivers for the SoC (UART, GPIO, PINMUX).

jfischer-phytec-iot commented:
SAM0 I2C driver was introduced by #14128, @madlain please rebase or close

galak commented:
Closing PR as its been marked stale for several months.  Feel free to re-open and update if desired.
